### DISCUSSION TITLE
Auswahllisten aus WebServices (DataLinq)

### DISCUSSION THREAD

**Discussion by marko-jer** [e-netze/webgis-community#30] [Ideas]
Pro Auswahlliste muss man momentan eine DataLinq-Query anlegen.
So entstehen viele Queries in einem Endpoint - oder man legt mehrere (viele) Endpoints an.

Für mich übersichtlicher wäre es, wenn mehrere Auswahllisten zum Beipeil zu einem Editierthema in einer Query sein könnten.
Analog zu den Datenbank-Queries könnte man die Auswahllisten mit "#if typ ... #endif" abgrenzen und auch so aufrufen.

**Comment by gis-eni** [comment]
Ja, du musst natürlich für jede Auswahlliste mit unterschiedlichen Werten eine eigene Abfrage im DataLinq anlegen. Das gleiche "Problem" hast du auch, wenn du für die Auswahlliste direkt über die Datenbank abholst.
Das ist allerdings auch so gewollt. Du kannst unter einem Endpoint beliebig viele Abfragen bauen. Genau dafür legt man einen Endpoint an. Mehrere Endpoints anzulegen (mit nur einer Abfrage) sollte man nicht. Ein Endpoint ist in gewisser Weise eine Klammer über Abfragen, die zusammen gehören. Das fördert die Übersicht ;)

Wie``#if / #endif`` hier aussehen sollte, erschließt sich mir nicht. Das solltest du hier vielleicht genauer mit einem Beispiel erklären.
Wenn ich mir das überlege, hab ich dann mehrere Auswahllisten in einer Abfrage?! Macht es das nicht erst recht kompliziert?

Eine Abfrage pro Auswahlliste: 
* Wenn man die Nomenklatur für Endpoints und Abfragen einmal für seine Anforderungen festlegt, ist das in de Regel nicht unübersichtlich
* Unabhängige Abfrage erhöhen die Stabilität: Packt man alle Auswahllisten in eine (komplexe) Abfrage mit ``#if / #endif`` läuft man immer Gefahr, mit einer Änderungen anderer Auswahllisten kaputt zu machen. Nach jeder Änderung müsste man somit immer alle Auswahllisten testen.

**Comment by marko-jer** [comment]
Ein Beispiel:
Ich habe ein EditierThema in dem mehrere Auswallisten verwendet werden. Da ist es übersichtlicher - für mich wenigstens, wenn diese Auswahllisten in einer Query/Abfrage gesammelt abrufbar sind.

Ich habe es nun so in einer Query/Abfrage gelöst, dass die einzelne Auswahlliste mit "level0" ausgewählt wird, die man über die "Db-Where Klausel" ansprechen kann. Das ersetzt somit das ` #if / #endif` .

Damit kann man die Anzahl der Queries/Abfragen reduzieren. 
Natürlich muss man auch mit dieser Methode eine Nomenklatur festlegen, damit es übersichtlich bleibt.

**Comment by gis-eni** [comment]
Ich kann mir nicht vorstellen, wie du mit ``level0`` mehrere Abfragen zusammengefasst hast. Kannst du hier bitte auch ein Beispiel posten?

**Comment by marko-jer** [comment]
Unter jedem `level0` ist eine Auswahlliste!

so schaut die Definition in der DataLinq-PlainText-Query aus:
```
vertragsstatus
  :
  MTB-Vertrag
  Wegefreigabeerklärung

typ-schwierigkeitsgrad
  :
  Trail
    :
    grün
    blau
    rot
    schwarz
  Route
    :
    blau
    rot
    schwarz

janein
  :
  ja:Ja
  nein:Nein
```
und so rufe ich die Auswahlliste in CMS auf:
![image](https://user-images.githubusercontent.com/1842500/215456790-8832a121-9cf9-4d81-95da-f197df9c1924.png)

**Comment by gis-eni** [comment]
Ok, kann man so machen.
In diesem Fall machst du aus drei Queries eine. Ist natürlich Geschmackssache. Ich empfehle allerdings nicht, das hier als "best pactice" zu nehmen. Einige Nachteile hab ich oben schon besprochen. Für "PlainText" abfragen gibt es noch folgende Nachteile:
* Die Abfrage wird einfach länger => scrollen notwendig
* Die Suche im DataLinq berücksichtig nur die Namen im Baum. Eine Suche nach einer Auswahlliste, zB "janein" (oder ``lov-janein``) wird keine Ergebnis bringen. Man muss immer wissen, in welcher "Abfrage" alles definiert ist
* Performance: Am Server muss immer beim Abholen der Werte ein großen Textfile geparst und gefiltert werden. Sollte man für Anwendungen mit vielen Zugriffen eher vermeiden. String-Manipulation ist "teuer"

**Comment by gis-eni** [comment]
PS:
**Zum ConnectionString:** Best Practice ist hier, die Connection inkl. Token (man sollte wenigstens den Endpoint schützen) in ein CMS Secret zu packen. Im CMS gibt man dann nur das Secret und die Query an:
``{{secret-datalinq-plaintext-lov}}@e-mtb-strecke`` an.
Secrets haben den Vorteil, dass man sie nicht immer neue eingeben muss. Außerdem muss eine Änderung am ConnectionString (zB neuer Token) nur an einer Stelle ausgebessert werden.

**Zu Queries:** Best Practice - Eine Abfrage pro Auswahlliste (Ausnahme: kaskadierende Abfragen)
Die ConnectionStrings im CMS würden dann so ausschauen:

``{{secret-datalinq-plaintext-lov}}@lov-e-mtb-strecke-vertragsstatus``,
``{{secret-datalinq-plaintext-lov}}@lov-e-mtb-strecke-typ-schwierigkeitsgrad``,
``{{secret-datalinq-plaintext-lov}}@lov-e-mtb-strecke-janein``

Nomenklatur:
``secret-datalinq`` --- ein Secret, dass sich auf Datalinq Quelle bezieht

``paintext-lov`` --- wäre hier beispielsweise der Name des Endpoints. Ein Endpoint könnte alle *Plaintext* Abfragen enthalten. Allerdings würde ich empfehlen, dass die Endpoint Namen im DataLinq mit dem Projekt beginnen, in dem die Abfragen und Views verwendet werden. Bei Auswahllisten fürs Editing wäre es beispielsweise gut, wenn der Endpoint mit dem Namen des *FeatureService* beginnen würde: ``radstrecken-lov``, wobei ``lov`` für *List Of Values* stehen könnte.

Die Query sollten für Auswahllisten mit ``lov`` beginnen oder enden. Ebenfalls sollte der Name des Editthemas (``e-mdb-strecke``) und das Feld (``vertragsstatus``, ``schwierigkeitsgrad``) darin vorkommen.

Hält man sich an diese Nomenklatur, finden man die einzelnen Abfragen sehr schnell im DataLinq.Code Baum. Gibt man in der Suche über dem Baum etwa ``vertragsstatus`` ein, klappt der Baum beim entsprechenden Knoten auf und man kann die Auswahllisten erweitern.
Schreibt man alle Auswahllisten in eine Query, findet man das so nicht mehr.
